### ISSUE TITLE
Support custom identity key for facts

### ISSUE THREAD

**Issue by douglasg14b** [NRules/NRules#336]
It seems that the `Update()` workflow is broken when `record class` are utilized. Likely because of the value-type semantics of records, which mean when a record is changed, the `HashCode` changes, making any keys in say a dictionary change (In ways they are not supposed to), making the fact un-findable.

Records should be treated as immutable, however, the `Update()` method doesn't seem to support this as it expects the same object instance to be mutated, not a new object instance returned (ie. with the non-destructive mutation feature)

- If I mutate a record with a different value in an `Update()` I receive a `Facts for update do not exist` exception.
- If I use `Update()` the `myRecord with { MyProp = newValue }` syntax, I appear to enter an infinite loop of some sort
- If I retract the fact, change the fact, and then insert the fact I end u in an infinite retract -> insert -> retract -> insert loop for the rule(s)

It also seems that because of the retraction of the fact in order to update it, it is not possible to make use of the `[Repeatability]` attribute 🤔 

If there is already support for this, how? If not, what sort of outlook does such support have? Is this something I could help implement?

**Comment by douglasg14b** [comment]
I'm browsing through the source of `Update()` `UpdateAll()` and `TryUpdateAll()`. I assume this is failing at the following line (returning null from the call): 

https://github.com/NRules/NRules/blob/b8c44a32bbbe0cc57edda0152004fd7d87374c54/src/NRules/NRules/Session.cs#L430

Which tried to retrieve this from a dictionary, with the key being the fact (The key that has now changed if the record was mutated). And thus it would probably fail to find the fact.

What side effects would there be from supporting an `original` and `modified` fact in a `TryUpdateAll()` overload? I imagine this would be more involved than just swapping out the value-like fact inside of `_factMap`?

<hr>

Side note, this is a bit of a pain without SourceLink (_wink-wink nudge-nudge_), can't quite tell what is exactly happening, just speculating 😛

Edit: Ignore my SourceLink poking, saw your other post. I can always build from source and add a nuget source to the output dir

**Comment by snikolayev** [comment]
@douglasg14b the thing to highlight here is that each fact is its own key in the working memory (i.e. it's looked up in a dictionary, like you pointed out), so regardless of something being a class or a record, the expectation is that the hash code won't change. In other words, facts are expected to have identity.
This is not a problem with records specifically, but immediately noticeable with mutable records due to value semantics.
While I personally believe mutable value objects (or mutable records with value semantics) are evil, I think there are other uses for this feature request, e.g. with classes where its equality comparison/hash code implementation are not suitable for use in the engine. So, I'm changing the title of this request to: allow inserting/updating facts with a custom identity key.
I think it's a reasonable feature request, so I'll look into it.

**Comment by douglasg14b** [comment]
Thanks for taking a look!

I think this might be capturing two things, and the title captures 1.

1. Classes with custom equality/hash code
   - Support here may be very difficult, it not outright impractical due to changing hashcodes if the object is mutated. Perhaps this is stated as a limitation in the same way that the base framework states these are limitations of dictionaries/hashsets...etc
3. Immutable types (class || struct) with custom equality/hash code
   - It seems more reasonable for this to be a support option. Being able to update a fact by passing in a replacement. This also supports the idea that records are intended to be used this way, and even have syntax specially made just for this purpose. It also avoids touching the demons that are changing hash codes.


Personally support for `#2` is more attractive, more reasonable, and doesn't cause any hardship to library users. Less about supporting custom identities, and more about supporting updating immutable facts without some the pitfalls of manually retracting & re-inserting them (ie. infinite loops).

Maybe I'm way off base, I'll let you comment.

This is something I'd be willing to put work in to help make happen, if that changes anything. I'd probably need some tech design & considerations from you.

<hr>

Edit: Unfortunately the project I was on which was making great use of N-Rules got canned because the tea was re-orged. I do not have a good use case for NRules at this time, and as such will not have the kindling needed to focus on it & contribute. 😞

**Comment by snikolayev** [comment]
Maybe there is a better way to phrase this, I'm looking at this as a way to pass an identity key for a fact where existing identity is not suitable (case 1) or there is no identity (case 2).
I don't think this is going to be a major change and should mostly stay within the existing design. When facts are passed to the session for insert/update, first thing it does is it looks up the Fact container in the dictionary. This lookup is done using the fact object itself as a key. Just need a way to pass a custom identity key. Once Fact container is obtained, the rest of the library will remain unchanged.
If you get to it first, feel free to take a shot at implementation. I'll try to get to this soon.

**Comment by DarkHarlock** [comment]
In the past few days, I had the chance to try out this library and encountered the same issue. So today, I tried to see how it would be possible to separate the concept of identity from that of equality without changing the API layout. Looking at the code, it seems the best strategy is to configure the `RuleCompiler` instance with a custom `IEqualityProvider<object>` so that every map used to locate the `Fact` instance from the containing object can have a user-side override.

By jotting down some code locally, I managed to get a series of working tests that leverage this pattern. If you think the idea is good, I can try to tidy up the code with some comments and create a PR.

Here is an example of usage taken from some tests I tried to add:

```cs
public class OneCustomComparerFactRuleTest : BaseRulesTestFixture
{
    /* ... */
    [Fact]
    public void Fire_DuplicateInsert_Throws()
    {
        //Arrange
        var fact1 = new FactType(1) { TestProperty = "Valid Value 1" };
        var fact2 = new FactType(1) { TestProperty = "Valid Value 2" };

        //Act - Assert
        Session.Insert(fact1);
        Assert.Throws<ArgumentException>(() => Session.Insert(fact2));
        Assert.NotEqual(fact1, fact2);
    }

    protected override void SetUpRules(IRulesTestSetup setup)
    {
        setup.RuleCompilerOptionFactory = () => new RuleCompilerOptions { FactComparer = new FactCustomEqualityComparer() };
        setup.Rule<TestRule>();
    }

    public interface IProvideFactIdentity
    {
        object Identity { get; }
    }

    public class FactCustomEqualityComparer : EqualityComparer<object>
    {
        public override bool Equals(object? x, object? y)
        {
            if (x is IProvideFactIdentity { } xIdentity && y is IProvideFactIdentity { } yIdentity)
            {
                (x, y) = (xIdentity.Identity, yIdentity.Identity);
            }
            return Default.Equals(x, y);
        }
        public override int GetHashCode(object obj)
        {
            obj = obj is IProvideFactIdentity identity ? identity.Identity : obj;
            return Default.GetHashCode(obj);
        }
    }

    public record class FactType(int Id) : IProvideFactIdentity
    {
        [NotNull]
        public string? TestProperty { get; init; }
        object IProvideFactIdentity.Identity => Id;
    }

    public class TestRule : Rule
    {
        public override void Define()
        {
            FactType fact = null!;

            When()
                .Match(() => fact, f => f.TestProperty.StartsWith("Valid"));
            Then()
                .Do(ctx => ctx.NoOp());
        }
    }
}
```

**Comment by snikolayev** [comment]
@DarkHarlock thank you for looking into this, and sorry for the late response.
This is certainly an interesting idea to have a central place to manage fact identity. I don't think it fully solves the problem with records, but I think it makes sense directionally and is a fine step forward. The reason I don't think it fully solves the problem, is this:
1. It will work for classes and reference type records, but if used with structs and struct records there is another layer of complexity around boxing and unboxing. Right now, NRules simply won't work with value types, but with the introduction of custom identity it may lead people to falsely believe it supports them, causing massive performance issues.
2. It will only allow the use of custom identities for facts managed through ISession. If application separates the runtime part from the rules part (i.e. rules exist in separate assemblies that don't have access to RuleCompiler), then they won't be able to use custom identity for any of the forward-chained facts defined on the rules side. This could be fixed by exposing an interface like IdentityProvider that you mentioned, but that's not ideal, as I firmly believe it is a bad practice to require domain objects to inherit from any framework-level base classes or implement any framework-level interfaces. So, I think this solution also needs another set of APIs where per-fact identity (or fact key) can be provided.

But as I said, I think this could be a good starting point, so if you are willing to create a PR and target a feature branch (instead of develop branch), such that we can collaborate on this, it would be really great and highly appreciated.

**Comment by snikolayev** [comment]
@DarkHarlock after playing with various ideas, I took your proposal as the foundation, added a few extensibility points, and implemented it in #371
